Release 4 preview (#294)
kata-containers/www.katacontainers.io: src/pages/blog/release-4-0-0-preview.md

  - Release 4.0.0 with runtime-rs as the default Kata runtime

# New Rust-based Default Runtime

The most significant architectural change in 4.0.0 is the switch to runtime-rs (the Rust-based runtime) as the default, while the original Go runtime will remain available but in a deprecated state.

## Background

There have been many advantages to have a project like Kata Containers implemented in Rust over Go. The community explored various areas to evaluate what users along with active contributors would gain by taking the leap, including:
- Reduced Memory Footprint: Essential for high-density container deployments.
- Improved Security: Leveraging Rust’s memory safety to further harden the container sandbox.
- Performance: Lower latency and faster startup times compared to its predecessor.

## Deprecation Period for the Go Runtime

While the Rust-based runtime will grab the spotlight, the Go runtime will still stick around for a deprecation period to support users as they switch from one to the other.

What you need to know:
- The Go runtime becomes deprecated when 4.0.0 is officially released
  - The Go runtime may be removed no sooner than the 5.0.0 release of Kata Containers

>     - 5.0.0 is currently forecasted to happen 18months after 4.0.0 is released

Reviewer: ```suggestion
    - 5.0.0 is currently forecast to happen 18 months after 4.0.0 is released
```

Reviewer: The 18months wasn't fixed?

Author: It is now, sorry about that.

  - No new feature development will be accepted in the Go runtime going forward, however, some exceptions apply, see below
  - The community will still accept and encourage bug and security fixes to the Go runtime
    - These fixes will have to be submitted for the Rust runtime as well, where applicable
- The community is considering a grace period to finish ongoing feature development work in the Go runtime, for up to 6 months after 4.0.0 is released
  - Features will be evaluated on a case-by-case basis
  - Features will need to be implemented in both the Go and Rust runtimes, where applicable

If you haven’t yet started looking into runtime-rs, the 4.0.0 release will be a great opportunity to not just start exploring it, but also start taking the steps to move your current deployment over to the new runtime. You can also reach out and work with the community if you identify any remaining gaps to close them.

## Highlights from the New Rust Runtime

> The community has been hard working to reach a close state to feature parity between the two variants of the runtime, where possible. Functionality that had already reached the state of deprecated or unmaintained will not be available in the new runtime, while it already has some features that are not available in the original Go runtime. This section gives a sneak peek into runtime-rs, without a comprehensive overview.

Reviewer: ```suggestion
The community has been working hard to reach a close state to feature parity between the two variants of the runtime, where possible. Functionality that had already reached the state of deprecated or unmaintained will not be available in the new runtime, while it already has some features that are not available in the original Go runtime. This section gives a sneak peek into runtime-rs, without a comprehensive overview.
```

### Hypervisor Support

The virtualization layer is one of the pillars of Kata Containers, and the project has been providing flexible options for users to use the VMM solution that best fits their needs.

runtime-rs will initially support the following hypervisors:
- QEMU
- Cloud Hypervisor
- Firecracker
- Dragonball - a special hypervisor created by the Ant Group that runs in the same process as the Rust-based containerd shim

### Hardware Architectures and Accelerators

It is important for the community to make the runtimes available on different hardware architectures. For runtime-rs these include:
- Arm
- Intel/AMD (x86_64)
- IBM s390x
- IBM PowerPC
- RISC-V - experimental in runtime-rs only

GPU support:
- Go runtime: GPU support tested with QEMU

> - Rust runtime: GPU support tested with QEMU

Reviewer: Do we need to comment on the go runtime here? If so maybe something like: `- Both runtimes support GPU passthrough with NVIDIA hardware when using QEMU` ?

  - The community will still accept and encourage bug and security fixes to the Go runtime
    - These fixes will have to be submitted for the Rust runtime as well, where applicable
- The community is considering a grace period to finish ongoing feature development work in the Go runtime, for up to 6 months after 4.0.0 is released
  - Features will be evaluated on a case-by-case basis
  - Features will need to be implemented in both the Go and Rust runtimes, where applicable

## Highlights from the New Rust Runtime

The community has been hard working to reach a close state to feature parity between the two variants of the runtime, where possible. Functionality that had already reached the state of deprecated or unmaintained will not be available in the new runtime, while some features have never gotten implemented in the original ones. This section gives a sneak peek into runtime-rs.

### Hypervisor Support

The virtualization layer is one of the pillars of Kata Containers, and the project has been providing flexible options for users to use the VMM solution that best fits their needs.

runtime-rs will initially support the following hypervisors:
- QEMU

> - Cloud Hypervisor?

Reviewer 2: We have CI for CH these days.
```suggestion
- Cloud Hypervisor
```

> - Firecracker?

Reviewer 3: With the rework done by @ananos, I think we can say Firecracker will be there.

- Dragonball

### Hardware Architectures and Accelerators

It is important for the community to make the runtimes available on different hardware architectures. For runtime-rs these include:
- Intel/AMD (x86_64)
- Arm
- RISC-V - experimental

> - ?

Reviewer 3: IBM s390x
IBM PowerPC

The original Go runtime already adds GPU support with a wide variety of configuration options. The initial GPU support in runtime-rs will include:

> - ?

Reviewer 3: The original Go runtime only has tested support for GPU with QEMU.

That will be on pair (as I already have a PR opened for that) for the runtime-rs.

### Container Orchestration System Compatibility

Kata Containers was designed to integrate seamlessly into existing container orchestration engines to become a vessel to package and manage workloads. This is the primary operating model for the Rust-based runtime as well, which will work with:
- Kubernetes

> - Docker?

Reviewer 3: Docker is okay, but we must explicitly say that it's only tested with runtime-rs + QEMU for now.

### Deploying runtime-rs

- From source
- kata-deploy

> - ?

Reviewer 3: We're removing kata-manager soon and kata-deploy will become the way to deploy for k8s and non-k8s. so nothing else to be added here.

> WHAT ELSE SHOULD WE HIGHLIGHT?

Reviewer 3: I think the addition of dragonball should be highlighted here, and then I'd ask @lifupan / @Apokleos to add more info.

Author: It's called out in the hypervisor section above, but I haven't found much information about it to provide more details. I would appreciate more info or pointers.
